The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7972417289/job/21764142550) [(plain)](https://github.com/rust-lang/rust/commit/c315461c78e03785922c370377828f9f6b16f317/checks/21764142550/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=ChrisDenton
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_723db8f0-201d-46c7-bac3-7b528ffd8dc2
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=ProcessPrng
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_723db8f0-201d-46c7-bac3-7b528ffd8dc2
GITHUB_REF=refs/pull/121337/merge
GITHUB_REF_NAME=121337/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=50db27ac861942291a6e1843bacaf25c4aa10f11
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_723db8f0-201d-46c7-bac3-7b528ffd8dc2
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_723db8f0-201d-46c7-bac3-7b528ffd8dc2
GITHUB_TRIGGERING_ACTOR=ChrisDenton
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/121337/merge
GITHUB_WORKFLOW_SHA=50db27ac861942291a6e1843bacaf25c4aa10f11
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
## Running ui tests in tests/pass against miri for i686-pc-windows-msvc
   Compiler: "MIRI_ENV_VAR_TEST"="0" "MIRI_TEMP"="/tmp" /checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/miri "--error-format=json" "-Dwarnings" "-Dunused" "-Ainternal_features" "-Zui-testing" "--target" "i686-pc-windows-msvc" "--out-dir" OUT_DIR

FAILED TEST: tests/pass/const-addrs.rs
command: MIRI_ENV_VAR_TEST="0" MIRI_TEMP="/tmp" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/miri" "--error-format=json" "-Dwarnings" "-Dunused" "-Ainternal_features" "-Zui-testing" "--target" "i686-pc-windows-msvc" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/tests/pass" "tests/pass/const-addrs.rs" "-Zinline-mir=no" "--edition" "2021"
   0: tests failed


Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
error: pass test got exit status: 1, but expected 0

error: actual output differed from expected
Execute `./miri test --bless` to update `tests/pass/const-addrs.stderr` to the actual output
Execute `./miri test --bless` to update `tests/pass/const-addrs.stderr` to the actual output
--- tests/pass/const-addrs.stderr
+++ <stderr output>
+error: unsupported operation: can't call foreign function `ProcessPrng` on OS `windows`
+  --> RUSTLIB/std/src/sys/pal/PLATFORM/rand.rs:LL:CC
+   |
+LL |     let ret = unsafe { c::ProcessPrng(&mut v as *mut _ as *mut u8, mem::size_of_val(&v)) };
+   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ can't call foreign function `ProcessPrng` on OS `windows`
+   = help: this is likely not a bug in the program; it indicates that the program performed an operation that the interpreter does not support
error: test failed, to rerun pass `--test compiletest`

Caused by:
Caused by:
  process didn't exit successfully: `/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/compiletest-f5946aa5fa5ac232 --quiet` (exit status: 1)
+   = note: BACKTRACE:
+   = note: inside `std::sys::pal::PLATFORM::rand::hashmap_random_keys` at RUSTLIB/std/src/sys/pal/PLATFORM/rand.rs:LL:CC
+   = note: inside `std::hash::RandomState::new::KEYS::__init` at RUSTLIB/std/src/hash/random.rs:LL:CC
+   = note: inside closure at RUSTLIB/std/src/sys/pal/PLATFORM/thread_local/os_local.rs:LL:CC
+   = note: inside `std::sys::pal::PLATFORM::thread_local::lazy::LazyKeyInner::<std::cell::Cell<(u64, u64)>>::initialize::<{closure@std::hash::RandomState::new::KEYS::__getit::{closure#0}}>` at RUSTLIB/std/src/sys/pal/PLATFORM/thread_local/mod.rs:LL:CC
+   = note: inside `std::thread::local_impl::Key::<std::cell::Cell<(u64, u64)>>::try_initialize::<{closure@std::hash::RandomState::new::KEYS::__getit::{closure#0}}>` at RUSTLIB/std/src/sys/pal/PLATFORM/thread_local/os_local.rs:LL:CC
+   = note: inside `std::thread::local_impl::Key::<std::cell::Cell<(u64, u64)>>::get::<{closure@std::hash::RandomState::new::KEYS::__getit::{closure#0}}>` at RUSTLIB/std/src/sys/pal/PLATFORM/thread_local/os_local.rs:LL:CC
+   = note: inside `std::hash::RandomState::new::KEYS::__getit` at RUSTLIB/std/src/sys/pal/PLATFORM/thread_local/os_local.rs:LL:CC
+   = note: inside `std::thread::LocalKey::<std::cell::Cell<(u64, u64)>>::try_with::<{closure@std::hash::RandomState::new::{closure#0}}, std::hash::RandomState>` at RUSTLIB/std/src/thread/local.rs:LL:CC
+   = note: inside `std::thread::LocalKey::<std::cell::Cell<(u64, u64)>>::with::<{closure@std::hash::RandomState::new::{closure#0}}, std::hash::RandomState>` at RUSTLIB/std/src/thread/local.rs:LL:CC
+   = note: inside `<std::hash::RandomState as std::default::Default>::default` at RUSTLIB/std/src/hash/random.rs:LL:CC
+   = note: inside `<std::hash::RandomState as std::default::Default>::default` at RUSTLIB/std/src/hash/random.rs:LL:CC
+   = note: inside `<hashbrown::map::HashMap<usize, (), std::hash::RandomState> as std::default::Default>::default` at CARGO_REGISTRY/.../map.rs:LL:CC
+   = note: inside `<hashbrown::set::HashSet<usize, std::hash::RandomState> as std::default::Default>::default` at CARGO_REGISTRY/.../set.rs:LL:CC
+   = note: inside `<std::collections::HashSet<usize> as std::default::Default>::default` at RUSTLIB/std/src/collections/hash/set.rs:LL:CC
+note: inside `main`
+  --> $DIR/const-addrs.rs:LL:CC
+   |
+LL |     let mut addrs = HashSet::new();
---
+


error: there were 1 unmatched diagnostics that occurred outside the testfile and had no pattern
    Error: unsupported operation: can't call foreign function `ProcessPrng` on OS `windows`
full stderr:
full stderr:
error: unsupported operation: can't call foreign function `ProcessPrng` on OS `windows`
##[error]  --> /checkout/library/std/src/sys/pal/windows/rand.rs:8:24
   |
LL |     let ret = unsafe { c::ProcessPrng(&mut v as *mut _ as *mut u8, mem::size_of_val(&v)) };
   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ can't call foreign function `ProcessPrng` on OS `windows`
   = help: this is likely not a bug in the program; it indicates that the program performed an operation that the interpreter does not support
   = note: BACKTRACE:
   = note: inside `std::sys::pal::windows::rand::hashmap_random_keys` at /checkout/library/std/src/sys/pal/windows/rand.rs:8:24: 8:89
   = note: inside `std::hash::RandomState::new::KEYS::__init` at /checkout/library/std/src/hash/random.rs:68:23: 68:49
   = note: inside `std::hash::RandomState::new::KEYS::__init` at /checkout/library/std/src/hash/random.rs:68:23: 68:49
   = note: inside closure at /checkout/library/std/src/sys/pal/common/thread_local/os_local.rs:70:25: 70:33
   = note: inside `std::sys::pal::common::thread_local::lazy::LazyKeyInner::<std::cell::Cell<(u64, u64)>>::initialize::<{closure@std::hash::RandomState::new::KEYS::__getit::{closure#0}}>` at /checkout/library/std/src/sys/pal/common/thread_local/mod.rs:54:25: 54:31
   = note: inside `std::thread::local_impl::Key::<std::cell::Cell<(u64, u64)>>::try_initialize::<{closure@std::hash::RandomState::new::KEYS::__getit::{closure#0}}>` at /checkout/library/std/src/sys/pal/common/thread_local/os_local.rs:159:23: 159:52
   = note: inside `std::thread::local_impl::Key::<std::cell::Cell<(u64, u64)>>::get::<{closure@std::hash::RandomState::new::KEYS::__getit::{closure#0}}>` at /checkout/library/std/src/sys/pal/common/thread_local/os_local.rs:127:18: 127:43
   = note: inside `std::hash::RandomState::new::KEYS::__getit` at /checkout/library/std/src/sys/pal/common/thread_local/os_local.rs:62:21: 71:23
   = note: inside `std::thread::LocalKey::<std::cell::Cell<(u64, u64)>>::try_with::<{closure@std::hash::RandomState::new::{closure#0}}, std::hash::RandomState>` at /checkout/library/std/src/thread/local.rs:285:32: 285:50
   = note: inside `std::thread::LocalKey::<std::cell::Cell<(u64, u64)>>::with::<{closure@std::hash::RandomState::new::{closure#0}}, std::hash::RandomState>` at /checkout/library/std/src/thread/local.rs:262:9: 262:25
   = note: inside `<std::hash::RandomState as std::default::Default>::default` at /checkout/library/std/src/hash/random.rs:152:9: 152:27
   = note: inside `<std::hash::RandomState as std::default::Default>::default` at /checkout/library/std/src/hash/random.rs:152:9: 152:27
   = note: inside `<hashbrown::map::HashMap<usize, (), std::hash::RandomState> as std::default::Default>::default` at /cargo/registry/src/index.crates.io-6f17d22bba15001f/hashbrown-0.14.3/src/map.rs:2228:30: 2228:48
   = note: inside `<hashbrown::set::HashSet<usize, std::hash::RandomState> as std::default::Default>::default` at /cargo/registry/src/index.crates.io-6f17d22bba15001f/hashbrown-0.14.3/src/set.rs:1404:18: 1404:36
   = note: inside `<std::collections::HashSet<usize> as std::default::Default>::default` at /checkout/library/std/src/collections/hash/set.rs:1114:25: 1114:43
note: inside `main`
  --> tests/pass/const-addrs.rs:16:21
   |
LL |     let mut addrs = HashSet::new();
---



FAILED TEST: tests/pass/float_nan.rs
command: MIRI_ENV_VAR_TEST="0" MIRI_TEMP="/tmp" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/miri" "--error-format=json" "-Dwarnings" "-Dunused" "-Ainternal_features" "-Zui-testing" "--target" "i686-pc-windows-msvc" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/tests/pass" "tests/pass/float_nan.rs" "--edition" "2021"
error: pass test got exit status: 1, but expected 0

error: actual output differed from expected
error: actual output differed from expected
Execute `./miri test --bless` to update `tests/pass/float_nan.stderr` to the actual output
--- tests/pass/float_nan.stderr
+++ <stderr output>
+error: unsupported operation: can't call foreign function `ProcessPrng` on OS `windows`
+  --> RUSTLIB/std/src/sys/pal/PLATFORM/rand.rs:LL:CC
+   |
+LL |     let ret = unsafe { c::ProcessPrng(&mut v as *mut _ as *mut u8, mem::size_of_val(&v)) };
+   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ can't call foreign function `ProcessPrng` on OS `windows`
+   = help: this is likely not a bug in the program; it indicates that the program performed an operation that the interpreter does not support
+   = note: BACKTRACE:
+   = note: inside `std::sys::pal::PLATFORM::rand::hashmap_random_keys` at RUSTLIB/std/src/sys/pal/PLATFORM/rand.rs:LL:CC
+   = note: inside `std::hash::RandomState::new::KEYS::__init` at RUSTLIB/std/src/hash/random.rs:LL:CC
+   = note: inside `std::hash::RandomState::new::KEYS::__init` at RUSTLIB/std/src/hash/random.rs:LL:CC
+   = note: inside closure at RUSTLIB/std/src/sys/pal/PLATFORM/thread_local/os_local.rs:LL:CC
+   = note: inside `std::sys::pal::PLATFORM::thread_local::lazy::LazyKeyInner::<std::cell::Cell<(u64, u64)>>::initialize::<{closure@std::hash::RandomState::new::KEYS::__getit::{closure#0}}>` at RUSTLIB/std/src/sys/pal/PLATFORM/thread_local/mod.rs:LL:CC
+   = note: inside `std::thread::local_impl::Key::<std::cell::Cell<(u64, u64)>>::try_initialize::<{closure@std::hash::RandomState::new::KEYS::__getit::{closure#0}}>` at RUSTLIB/std/src/sys/pal/PLATFORM/thread_local/os_local.rs:LL:CC
+   = note: inside `std::thread::local_impl::Key::<std::cell::Cell<(u64, u64)>>::get::<{closure@std::hash::RandomState::new::KEYS::__getit::{closure#0}}>` at RUSTLIB/std/src/sys/pal/PLATFORM/thread_local/os_local.rs:LL:CC
+   = note: inside `std::hash::RandomState::new::KEYS::__getit` at RUSTLIB/std/src/sys/pal/PLATFORM/thread_local/os_local.rs:LL:CC
+   = note: inside `std::thread::LocalKey::<std::cell::Cell<(u64, u64)>>::try_with::<{closure@std::hash::RandomState::new::{closure#0}}, std::hash::RandomState>` at RUSTLIB/std/src/thread/local.rs:LL:CC
+   = note: inside `std::thread::LocalKey::<std::cell::Cell<(u64, u64)>>::with::<{closure@std::hash::RandomState::new::{closure#0}}, std::hash::RandomState>` at RUSTLIB/std/src/thread/local.rs:LL:CC
+   = note: inside `<std::hash::RandomState as std::default::Default>::default` at RUSTLIB/std/src/hash/random.rs:LL:CC
+   = note: inside `<std::hash::RandomState as std::default::Default>::default` at RUSTLIB/std/src/hash/random.rs:LL:CC
+   = note: inside `<std::collections::HashSet<F32> as std::iter::FromIterator<F32>>::from_iter::<[F32; 2]>` at RUSTLIB/std/src/collections/hash/set.rs:LL:CC
+note: inside `test_f32`
+   |
+   |
+LL |         HashSet::from_iter([F32::nan(Pos, Quiet, 0), F32::nan(Neg, Quiet, 0)]),
+note: inside `main`
+  --> $DIR/float_nan.rs:LL:CC
+   |
+LL |     test_f32();
---
+


error: there were 1 unmatched diagnostics that occurred outside the testfile and had no pattern
    Error: unsupported operation: can't call foreign function `ProcessPrng` on OS `windows`
full stderr:
full stderr:
error: unsupported operation: can't call foreign function `ProcessPrng` on OS `windows`
##[error]  --> /checkout/library/std/src/sys/pal/windows/rand.rs:8:24
   |
LL |     let ret = unsafe { c::ProcessPrng(&mut v as *mut _ as *mut u8, mem::size_of_val(&v)) };
   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ can't call foreign function `ProcessPrng` on OS `windows`
   = help: this is likely not a bug in the program; it indicates that the program performed an operation that the interpreter does not support
   = note: BACKTRACE:
   = note: inside `std::sys::pal::windows::rand::hashmap_random_keys` at /checkout/library/std/src/sys/pal/windows/rand.rs:8:24: 8:89
   = note: inside `std::hash::RandomState::new::KEYS::__init` at /checkout/library/std/src/hash/random.rs:68:23: 68:49
   = note: inside `std::hash::RandomState::new::KEYS::__init` at /checkout/library/std/src/hash/random.rs:68:23: 68:49
   = note: inside closure at /checkout/library/std/src/sys/pal/common/thread_local/os_local.rs:70:25: 70:33
   = note: inside `std::sys::pal::common::thread_local::lazy::LazyKeyInner::<std::cell::Cell<(u64, u64)>>::initialize::<{closure@std::hash::RandomState::new::KEYS::__getit::{closure#0}}>` at /checkout/library/std/src/sys/pal/common/thread_local/mod.rs:54:25: 54:31
   = note: inside `std::thread::local_impl::Key::<std::cell::Cell<(u64, u64)>>::try_initialize::<{closure@std::hash::RandomState::new::KEYS::__getit::{closure#0}}>` at /checkout/library/std/src/sys/pal/common/thread_local/os_local.rs:159:23: 159:52
   = note: inside `std::thread::local_impl::Key::<std::cell::Cell<(u64, u64)>>::get::<{closure@std::hash::RandomState::new::KEYS::__getit::{closure#0}}>` at /checkout/library/std/src/sys/pal/common/thread_local/os_local.rs:127:18: 127:43
   = note: inside `std::hash::RandomState::new::KEYS::__getit` at /checkout/library/std/src/sys/pal/common/thread_local/os_local.rs:62:21: 71:23
   = note: inside `std::thread::LocalKey::<std::cell::Cell<(u64, u64)>>::try_with::<{closure@std::hash::RandomState::new::{closure#0}}, std::hash::RandomState>` at /checkout/library/std/src/thread/local.rs:285:32: 285:50
   = note: inside `std::thread::LocalKey::<std::cell::Cell<(u64, u64)>>::with::<{closure@std::hash::RandomState::new::{closure#0}}, std::hash::RandomState>` at /checkout/library/std/src/thread/local.rs:262:9: 262:25
   = note: inside `<std::hash::RandomState as std::default::Default>::default` at /checkout/library/std/src/hash/random.rs:152:9: 152:27
   = note: inside `<std::hash::RandomState as std::default::Default>::default` at /checkout/library/std/src/hash/random.rs:152:9: 152:27
   = note: inside `<std::collections::HashSet<F32> as std::iter::FromIterator<F32>>::from_iter::<[F32; 2]>` at /checkout/library/std/src/collections/hash/set.rs:1025:44: 1025:62
note: inside `test_f32`
   |
   |
LL |         HashSet::from_iter([F32::nan(Pos, Quiet, 0), F32::nan(Neg, Quiet, 0)]),
note: inside `main`
  --> tests/pass/float_nan.rs:587:5
   |
LL |     test_f32();
---



FAILED TEST: tests/pass/hashmap.rs (revision `stack`)
command: MIRI_ENV_VAR_TEST="0" MIRI_TEMP="/tmp" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/miri" "--error-format=json" "-Dwarnings" "-Dunused" "-Ainternal_features" "-Zui-testing" "--target" "i686-pc-windows-msvc" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/tests/pass" "tests/pass/hashmap.rs" "--cfg=stack" "--edition" "2021"
error: pass test got exit status: 1, but expected 0

error: actual output differed from expected
Execute `./miri test --bless` to update `tests/pass/hashmap.stack.stderr` to the actual output
Execute `./miri test --bless` to update `tests/pass/hashmap.stack.stderr` to the actual output
--- tests/pass/hashmap.stack.stderr
+++ <stderr output>
+error: unsupported operation: can't call foreign function `ProcessPrng` on OS `windows`
+  --> RUSTLIB/std/src/sys/pal/PLATFORM/rand.rs:LL:CC
+   |
+LL |     let ret = unsafe { c::ProcessPrng(&mut v as *mut _ as *mut u8, mem::size_of_val(&v)) };
+   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ can't call foreign function `ProcessPrng` on OS `windows`
+   = help: this is likely not a bug in the program; it indicates that the program performed an operation that the interpreter does not support
+   = note: BACKTRACE:
+   = note: inside `std::sys::pal::PLATFORM::rand::hashmap_random_keys` at RUSTLIB/std/src/sys/pal/PLATFORM/rand.rs:LL:CC
+   = note: inside `std::hash::RandomState::new::KEYS::__init` at RUSTLIB/std/src/hash/random.rs:LL:CC
+   = note: inside `std::hash::RandomState::new::KEYS::__init` at RUSTLIB/std/src/hash/random.rs:LL:CC
+   = note: inside closure at RUSTLIB/std/src/sys/pal/PLATFORM/thread_local/os_local.rs:LL:CC
+   = note: inside `std::sys::pal::PLATFORM::thread_local::lazy::LazyKeyInner::<std::cell::Cell<(u64, u64)>>::initialize::<{closure@std::hash::RandomState::new::KEYS::__getit::{closure#0}}>` at RUSTLIB/std/src/sys/pal/PLATFORM/thread_local/mod.rs:LL:CC
+   = note: inside `std::thread::local_impl::Key::<std::cell::Cell<(u64, u64)>>::try_initialize::<{closure@std::hash::RandomState::new::KEYS::__getit::{closure#0}}>` at RUSTLIB/std/src/sys/pal/PLATFORM/thread_local/os_local.rs:LL:CC
+   = note: inside `std::thread::local_impl::Key::<std::cell::Cell<(u64, u64)>>::get::<{closure@std::hash::RandomState::new::KEYS::__getit::{closure#0}}>` at RUSTLIB/std/src/sys/pal/PLATFORM/thread_local/os_local.rs:LL:CC
+   = note: inside `std::hash::RandomState::new::KEYS::__getit` at RUSTLIB/std/src/sys/pal/PLATFORM/thread_local/os_local.rs:LL:CC
+   = note: inside `std::thread::LocalKey::<std::cell::Cell<(u64, u64)>>::try_with::<{closure@std::hash::RandomState::new::{closure#0}}, std::hash::RandomState>` at RUSTLIB/std/src/thread/local.rs:LL:CC
+   = note: inside `std::thread::LocalKey::<std::cell::Cell<(u64, u64)>>::with::<{closure@std::hash::RandomState::new::{closure#0}}, std::hash::RandomState>` at RUSTLIB/std/src/thread/local.rs:LL:CC
+   = note: inside `<std::hash::RandomState as std::default::Default>::default` at RUSTLIB/std/src/hash/random.rs:LL:CC
+   = note: inside `<std::collections::HashMap<i32, i32> as std::default::Default>::default` at RUSTLIB/std/src/collections/hash/map.rs:LL:CC
+   = note: inside `std::collections::HashMap::<i32, i32>::new` at RUSTLIB/std/src/collections/hash/map.rs:LL:CC
+note: inside `main`
---
+


error: there were 1 unmatched diagnostics that occurred outside the testfile and had no pattern
    Error: unsupported operation: can't call foreign function `ProcessPrng` on OS `windows`
full stderr:
full stderr:
error: unsupported operation: can't call foreign function `ProcessPrng` on OS `windows`
##[error]  --> /checkout/library/std/src/sys/pal/windows/rand.rs:8:24
   |
LL |     let ret = unsafe { c::ProcessPrng(&mut v as *mut _ as *mut u8, mem::size_of_val(&v)) };
   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ can't call foreign function `ProcessPrng` on OS `windows`
   = help: this is likely not a bug in the program; it indicates that the program performed an operation that the interpreter does not support
   = note: BACKTRACE:
   = note: inside `std::sys::pal::windows::rand::hashmap_random_keys` at /checkout/library/std/src/sys/pal/windows/rand.rs:8:24: 8:89
   = note: inside `std::hash::RandomState::new::KEYS::__init` at /checkout/library/std/src/hash/random.rs:68:23: 68:49
   = note: inside `std::hash::RandomState::new::KEYS::__init` at /checkout/library/std/src/hash/random.rs:68:23: 68:49
   = note: inside closure at /checkout/library/std/src/sys/pal/common/thread_local/os_local.rs:70:25: 70:33
   = note: inside `std::sys::pal::common::thread_local::lazy::LazyKeyInner::<std::cell::Cell<(u64, u64)>>::initialize::<{closure@std::hash::RandomState::new::KEYS::__getit::{closure#0}}>` at /checkout/library/std/src/sys/pal/common/thread_local/mod.rs:54:25: 54:31
   = note: inside `std::thread::local_impl::Key::<std::cell::Cell<(u64, u64)>>::try_initialize::<{closure@std::hash::RandomState::new::KEYS::__getit::{closure#0}}>` at /checkout/library/std/src/sys/pal/common/thread_local/os_local.rs:159:23: 159:52
   = note: inside `std::thread::local_impl::Key::<std::cell::Cell<(u64, u64)>>::get::<{closure@std::hash::RandomState::new::KEYS::__getit::{closure#0}}>` at /checkout/library/std/src/sys/pal/common/thread_local/os_local.rs:127:18: 127:43
   = note: inside `std::hash::RandomState::new::KEYS::__getit` at /checkout/library/std/src/sys/pal/common/thread_local/os_local.rs:62:21: 71:23
   = note: inside `std::thread::LocalKey::<std::cell::Cell<(u64, u64)>>::try_with::<{closure@std::hash::RandomState::new::{closure#0}}, std::hash::RandomState>` at /checkout/library/std/src/thread/local.rs:285:32: 285:50
   = note: inside `std::thread::LocalKey::<std::cell::Cell<(u64, u64)>>::with::<{closure@std::hash::RandomState::new::{closure#0}}, std::hash::RandomState>` at /checkout/library/std/src/thread/local.rs:262:9: 262:25
   = note: inside `<std::hash::RandomState as std::default::Default>::default` at /checkout/library/std/src/hash/random.rs:152:9: 152:27
   = note: inside `<std::collections::HashMap<i32, i32> as std::default::Default>::default` at /checkout/library/std/src/collections/hash/map.rs:1321:30: 1321:48
   = note: inside `std::collections::HashMap::<i32, i32>::new` at /checkout/library/std/src/collections/hash/map.rs:233:9: 233:27
note: inside `main`
---



FAILED TEST: tests/pass/hashmap.rs (revision `tree`)
command: MIRI_ENV_VAR_TEST="0" MIRI_TEMP="/tmp" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/miri" "--error-format=json" "-Dwarnings" "-Dunused" "-Ainternal_features" "-Zui-testing" "--target" "i686-pc-windows-msvc" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/tests/pass" "tests/pass/hashmap.rs" "--cfg=tree" "-Zmiri-tree-borrows" "--edition" "2021"
error: pass test got exit status: 1, but expected 0

error: actual output differed from expected
error: actual output differed from expected
Execute `./miri test --bless` to update `tests/pass/hashmap.tree.stderr` to the actual output
--- tests/pass/hashmap.tree.stderr
+++ <stderr output>
+error: unsupported operation: can't call foreign function `ProcessPrng` on OS `windows`
+  --> RUSTLIB/std/src/sys/pal/PLATFORM/rand.rs:LL:CC
+   |
+LL |     let ret = unsafe { c::ProcessPrng(&mut v as *mut _ as *mut u8, mem::size_of_val(&v)) };
+   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ can't call foreign function `ProcessPrng` on OS `windows`
+   = help: this is likely not a bug in the program; it indicates that the program performed an operation that the interpreter does not support
+   = note: BACKTRACE:
+   = note: inside `std::sys::pal::PLATFORM::rand::hashmap_random_keys` at RUSTLIB/std/src/sys/pal/PLATFORM/rand.rs:LL:CC
+   = note: inside `std::hash::RandomState::new::KEYS::__init` at RUSTLIB/std/src/hash/random.rs:LL:CC
+   = note: inside `std::hash::RandomState::new::KEYS::__init` at RUSTLIB/std/src/hash/random.rs:LL:CC
+   = note: inside closure at RUSTLIB/std/src/sys/pal/PLATFORM/thread_local/os_local.rs:LL:CC
+   = note: inside `std::sys::pal::PLATFORM::thread_local::lazy::LazyKeyInner::<std::cell::Cell<(u64, u64)>>::initialize::<{closure@std::hash::RandomState::new::KEYS::__getit::{closure#0}}>` at RUSTLIB/std/src/sys/pal/PLATFORM/thread_local/mod.rs:LL:CC
+   = note: inside `std::thread::local_impl::Key::<std::cell::Cell<(u64, u64)>>::try_initialize::<{closure@std::hash::RandomState::new::KEYS::__getit::{closure#0}}>` at RUSTLIB/std/src/sys/pal/PLATFORM/thread_local/os_local.rs:LL:CC
+   = note: inside `std::thread::local_impl::Key::<std::cell::Cell<(u64, u64)>>::get::<{closure@std::hash::RandomState::new::KEYS::__getit::{closure#0}}>` at RUSTLIB/std/src/sys/pal/PLATFORM/thread_local/os_local.rs:LL:CC
+   = note: inside `std::hash::RandomState::new::KEYS::__getit` at RUSTLIB/std/src/sys/pal/PLATFORM/thread_local/os_local.rs:LL:CC
+   = note: inside `std::thread::LocalKey::<std::cell::Cell<(u64, u64)>>::try_with::<{closure@std::hash::RandomState::new::{closure#0}}, std::hash::RandomState>` at RUSTLIB/std/src/thread/local.rs:LL:CC
+   = note: inside `std::thread::LocalKey::<std::cell::Cell<(u64, u64)>>::with::<{closure@std::hash::RandomState::new::{closure#0}}, std::hash::RandomState>` at RUSTLIB/std/src/thread/local.rs:LL:CC
+   = note: inside `<std::hash::RandomState as std::default::Default>::default` at RUSTLIB/std/src/hash/random.rs:LL:CC
+   = note: inside `<std::collections::HashMap<i32, i32> as std::default::Default>::default` at RUSTLIB/std/src/collections/hash/map.rs:LL:CC
+   = note: inside `std::collections::HashMap::<i32, i32>::new` at RUSTLIB/std/src/collections/hash/map.rs:LL:CC
+note: inside `main`
---
+


error: there were 1 unmatched diagnostics that occurred outside the testfile and had no pattern
    Error: unsupported operation: can't call foreign function `ProcessPrng` on OS `windows`
full stderr:
full stderr:
error: unsupported operation: can't call foreign function `ProcessPrng` on OS `windows`
##[error]  --> /checkout/library/std/src/sys/pal/windows/rand.rs:8:24
   |
LL |     let ret = unsafe { c::ProcessPrng(&mut v as *mut _ as *mut u8, mem::size_of_val(&v)) };
   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ can't call foreign function `ProcessPrng` on OS `windows`
   = help: this is likely not a bug in the program; it indicates that the program performed an operation that the interpreter does not support
   = note: BACKTRACE:
   = note: inside `std::sys::pal::windows::rand::hashmap_random_keys` at /checkout/library/std/src/sys/pal/windows/rand.rs:8:24: 8:89
   = note: inside `std::hash::RandomState::new::KEYS::__init` at /checkout/library/std/src/hash/random.rs:68:23: 68:49
   = note: inside `std::hash::RandomState::new::KEYS::__init` at /checkout/library/std/src/hash/random.rs:68:23: 68:49
   = note: inside closure at /checkout/library/std/src/sys/pal/common/thread_local/os_local.rs:70:25: 70:33
   = note: inside `std::sys::pal::common::thread_local::lazy::LazyKeyInner::<std::cell::Cell<(u64, u64)>>::initialize::<{closure@std::hash::RandomState::new::KEYS::__getit::{closure#0}}>` at /checkout/library/std/src/sys/pal/common/thread_local/mod.rs:54:25: 54:31
   = note: inside `std::thread::local_impl::Key::<std::cell::Cell<(u64, u64)>>::try_initialize::<{closure@std::hash::RandomState::new::KEYS::__getit::{closure#0}}>` at /checkout/library/std/src/sys/pal/common/thread_local/os_local.rs:159:23: 159:52
   = note: inside `std::thread::local_impl::Key::<std::cell::Cell<(u64, u64)>>::get::<{closure@std::hash::RandomState::new::KEYS::__getit::{closure#0}}>` at /checkout/library/std/src/sys/pal/common/thread_local/os_local.rs:127:18: 127:43
   = note: inside `std::hash::RandomState::new::KEYS::__getit` at /checkout/library/std/src/sys/pal/common/thread_local/os_local.rs:62:21: 71:23
   = note: inside `std::thread::LocalKey::<std::cell::Cell<(u64, u64)>>::try_with::<{closure@std::hash::RandomState::new::{closure#0}}, std::hash::RandomState>` at /checkout/library/std/src/thread/local.rs:285:32: 285:50
   = note: inside `std::thread::LocalKey::<std::cell::Cell<(u64, u64)>>::with::<{closure@std::hash::RandomState::new::{closure#0}}, std::hash::RandomState>` at /checkout/library/std/src/thread/local.rs:262:9: 262:25
   = note: inside `<std::hash::RandomState as std::default::Default>::default` at /checkout/library/std/src/hash/random.rs:152:9: 152:27
   = note: inside `<std::collections::HashMap<i32, i32> as std::default::Default>::default` at /checkout/library/std/src/collections/hash/map.rs:1321:30: 1321:48
   = note: inside `std::collections::HashMap::<i32, i32>::new` at /checkout/library/std/src/collections/hash/map.rs:233:9: 233:27
note: inside `main`
```

</details>

--- a/compiler/rustc_codegen_cranelift/patches/0029-stdlib-rawdylib-processprng.patch
+++ b/compiler/rustc_codegen_cranelift/patches/0029-stdlib-rawdylib-processprng.patch
@@ -1,0 +1,47 @@
+From 9f65e742ba3e41474e6126c6c4469c48eaa6ca7e Mon Sep 17 00:00:00 2001
+From: Chris Denton <chris@chrisdenton.dev>
+Date: Tue, 20 Feb 2024 16:01:40 -0300
+Subject: [PATCH] Don't use raw-dylib in std
+
+---
+ library/std/src/sys/pal/windows/c.rs    | 2 +-
+ library/std/src/sys/pal/windows/rand.rs | 3 +--
+ 2 files changed, 2 insertions(+), 3 deletions(-)
+
+diff --git a/library/std/src/sys/pal/windows/c.rs b/library/std/src/sys/pal/windows/c.rs
+index ad8e01bfa9b..9ca8e4c16ce 100644
+--- a/library/std/src/sys/pal/windows/c.rs
++++ b/library/std/src/sys/pal/windows/c.rs
+@@ -323,7 +323,7 @@ pub unsafe fn NtWriteFile(
+ 
+ // Use raw-dylib to import ProcessPrng as we can't rely on there being an import library.
+ cfg_if::cfg_if! {
+-if #[cfg(not(target_vendor = "win7"))] {
++if #[cfg(any())] {
+     #[cfg(target_arch = "x86")]
+     #[link(name = "bcryptprimitives", kind = "raw-dylib", import_name_type = "undecorated")]
+     extern "system" {
+diff --git a/library/std/src/sys/pal/windows/rand.rs b/library/std/src/sys/pal/windows/rand.rs
+index e427546222a..f2fe42a4d51 100644
+--- a/library/std/src/sys/pal/windows/rand.rs
++++ b/library/std/src/sys/pal/windows/rand.rs
+@@ -2,7 +2,7 @@
+ use core::mem;
+ use core::ptr;
+ 
+-#[cfg(not(target_vendor = "win7"))]
++#[cfg(any())]
+ #[inline]
+ pub fn hashmap_random_keys() -> (u64, u64) {
+     let mut v = (0, 0);
+@@ -13,7 +13,6 @@ pub fn hashmap_random_keys() -> (u64, u64) {
+     v
+ }
+ 
+-#[cfg(target_vendor = "win7")]
+ pub fn hashmap_random_keys() -> (u64, u64) {
+     use crate::ffi::c_void;
+     use crate::io;
+-- 
+2.42.0.windows.2
+

--- a/library/std/src/sys/pal/windows/c.rs
+++ b/library/std/src/sys/pal/windows/c.rs
@@ -321,6 +321,21 @@ pub unsafe fn NtWriteFile(
 }
 }
 
+// Use raw-dylib to import ProcessPrng as we can't rely on there being an import library.
+cfg_if::cfg_if! {
+if #[cfg(not(target_vendor = "win7"))] {
+    #[cfg(target_arch = "x86")]
+    #[link(name = "bcryptprimitives", kind = "raw-dylib", import_name_type = "undecorated")]
+    extern "system" {
+        pub fn ProcessPrng(pbdata: *mut u8, cbdata: usize) -> BOOL;
+    }
+    #[cfg(not(target_arch = "x86"))]
+    #[link(name = "bcryptprimitives", kind = "raw-dylib")]
+    extern "system" {
+        pub fn ProcessPrng(pbdata: *mut u8, cbdata: usize) -> BOOL;
+    }
+}}
+
 // Functions that aren't available on every version of Windows that we support,
 // but we still use them and just provide some form of a fallback implementation.
 compat_fn_with_fallback! {

--- a/library/std/src/sys/pal/windows/c/bindings.txt
+++ b/library/std/src/sys/pal/windows/c/bindings.txt
@@ -2180,10 +2180,6 @@ Windows.Win32.Networking.WinSock.WSATRY_AGAIN
 Windows.Win32.Networking.WinSock.WSATYPE_NOT_FOUND
 Windows.Win32.Networking.WinSock.WSAVERNOTSUPPORTED
 Windows.Win32.Security.Authentication.Identity.RtlGenRandom
-Windows.Win32.Security.Cryptography.BCRYPT_ALG_HANDLE
-Windows.Win32.Security.Cryptography.BCRYPT_USE_SYSTEM_PREFERRED_RNG
-Windows.Win32.Security.Cryptography.BCryptGenRandom
-Windows.Win32.Security.Cryptography.BCRYPTGENRANDOM_FLAGS
 Windows.Win32.Security.SECURITY_ATTRIBUTES
 Windows.Win32.Security.TOKEN_ACCESS_MASK
 Windows.Win32.Security.TOKEN_ACCESS_PSEUDO_HANDLE

--- a/library/std/src/sys/pal/windows/c/windows_sys.rs
+++ b/library/std/src/sys/pal/windows/c/windows_sys.rs
@@ -15,15 +15,6 @@ extern "system" {
     pub fn RtlGenRandom(randombuffer: *mut ::core::ffi::c_void, randombufferlength: u32)
     -> BOOLEAN;
 }
-#[link(name = "bcrypt")]
-extern "system" {
-    pub fn BCryptGenRandom(
-        halgorithm: BCRYPT_ALG_HANDLE,
-        pbbuffer: *mut u8,
-        cbbuffer: u32,
-        dwflags: BCRYPTGENRANDOM_FLAGS,
-    ) -> NTSTATUS;
-}
 #[link(name = "kernel32")]
 extern "system" {
     pub fn AcquireSRWLockExclusive(srwlock: *mut SRWLOCK) -> ();
@@ -889,9 +880,6 @@ impl ::core::clone::Clone for ARM64_NT_NEON128_0 {
         *self
     }
 }
-pub type BCRYPTGENRANDOM_FLAGS = u32;
-pub type BCRYPT_ALG_HANDLE = *mut ::core::ffi::c_void;
-pub const BCRYPT_USE_SYSTEM_PREFERRED_RNG: BCRYPTGENRANDOM_FLAGS = 2u32;
 pub const BELOW_NORMAL_PRIORITY_CLASS: PROCESS_CREATION_FLAGS = 16384u32;
 pub type BOOL = i32;
 pub type BOOLEAN = u8;

--- a/library/std/src/sys/pal/windows/rand.rs
+++ b/library/std/src/sys/pal/windows/rand.rs
@@ -1,42 +1,27 @@
-use crate::mem;
-use crate::ptr;
 use crate::sys::c;
+use core::mem;
+use core::ptr;
 
+#[cfg(not(target_vendor = "win7"))]
+#[inline]
 pub fn hashmap_random_keys() -> (u64, u64) {
     let mut v = (0, 0);
-    let ret = unsafe {
-        c::BCryptGenRandom(
-            ptr::null_mut(),
-            core::ptr::addr_of_mut!(v) as *mut u8,
-            mem::size_of_val(&v) as c::ULONG,
-            c::BCRYPT_USE_SYSTEM_PREFERRED_RNG,
-        )
-    };
-    if c::nt_success(ret) { v } else { fallback_rng() }
+    let ret = unsafe { c::ProcessPrng(ptr::addr_of_mut!(v).cast::<u8>(), mem::size_of_val(&v)) };
+    // ProcessPrng is documented as always returning `TRUE`.
+    // https://learn.microsoft.com/en-us/windows/win32/seccng/processprng#return-value
+    debug_assert_eq!(ret, c::TRUE);
+    v
 }
 
-/// Generate random numbers using the fallback RNG function (RtlGenRandom)
-///
-/// This is necessary because of a failure to load the SysWOW64 variant of the
-/// bcryptprimitives.dll library from code that lives in bcrypt.dll
-/// See <https://bugzilla.mozilla.org/show_bug.cgi?id=1788004#c9>
-#[cfg(not(target_vendor = "uwp"))]
-#[inline(never)]
-fn fallback_rng() -> (u64, u64) {
+#[cfg(target_vendor = "win7")]
+pub fn hashmap_random_keys() -> (u64, u64) {
     use crate::ffi::c_void;
     use crate::io;
 
     let mut v = (0, 0);
     let ret = unsafe {
-        c::RtlGenRandom(core::ptr::addr_of_mut!(v) as *mut c_void, mem::size_of_val(&v) as c::ULONG)
+        c::RtlGenRandom(ptr::addr_of_mut!(v).cast::<c_void>(), mem::size_of_val(&v) as c::ULONG)
     };
 
-    if ret != 0 { v } else { panic!("fallback RNG broken: {}", io::Error::last_os_error()) }
-}
-
-/// We can't use RtlGenRandom with UWP, so there is no fallback
-#[cfg(target_vendor = "uwp")]
-#[inline(never)]
-fn fallback_rng() -> (u64, u64) {
-    panic!("fallback RNG broken: RtlGenRandom() not supported on UWP");
+    if ret != 0 { v } else { panic!("RNG broken: {}", io::Error::last_os_error()) }
 }

--- a/src/tools/miri/src/shims/windows/foreign_items.rs
+++ b/src/tools/miri/src/shims/windows/foreign_items.rs
@@ -427,6 +427,14 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
                 this.gen_random(ptr, len.into())?;
                 this.write_scalar(Scalar::from_bool(true), dest)?;
             }
+            "ProcessPrng" => {
+                let [ptr, len] =
+                    this.check_shim(abi, Abi::System { unwind: false }, link_name, args)?;
+                let ptr = this.read_pointer(ptr)?;
+                let len = this.read_target_usize(len)?;
+                this.gen_random(ptr, len.into())?;
+                this.write_scalar(Scalar::from_i32(1), dest)?;
+            }
             "BCryptGenRandom" => {
                 let [algorithm, ptr, len, flags] =
                     this.check_shim(abi, Abi::System { unwind: false }, link_name, args)?;

